File: 8-16.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(vector<vector<int>>& A) {
    vector<vector<int>> D(6, vector<int>(2, 0));
    D[0][0] = D[0][1] = A[0][1];
    for (int i = 1; i < 6; i++) {
        D[i][0] = min(D[i - 1][0] + A[2 * i - 2][2 * i + 1] + A[2 * i][2 * i + 1], \
            D[i - 1][1] + A[2 * i - 1][2 * i + 1] + A[2 * i][2 * i + 1]);
        D[i][1] = min(D[i - 1][0] + A[2 * i - 2][2 * i] + A[2 * i][2 * i + 1], \
            D[i - 1][1] + A[2 * i - 1][2 * i] + A[2 * i][2 * i + 1]);
    }
    return min(D[5][0], D[5][1]);
}

int main() {
    vector<vector<int>> A(12, vector<int>(12, 0));
    for (int i = 0; i < 12; i++) {
        for (int j = 0; j < 12; j++) {
            cin >> A[i][j];
        }
    }
    cout << solution(A) << endl;
}

File: 8-16.py
import sys

input = sys.stdin.readline


def solution(A):
    D = list([0] * 2 for _ in range(6))
    D[0][0] = D[0][1] = A[0][1]
    for i in range(1, 6):
        D[i][0] = min(D[i - 1][0] + A[2 * i - 2][2 * i + 1] + A[2 * i][2 * i + 1], \
                      D[i - 1][1] + A[2 * i - 1][2 * i + 1] + A[2 * i][2 * i + 1])

        D[i][1] = min(D[i - 1][0] + A[2 * i - 2][2 * i] + A[2 * i][2 * i + 1], \
                      D[i - 1][1] + A[2 * i - 1][2 * i] + A[2 * i][2 * i + 1])

    return min(D[5][0], D[5][1])


A = list(list(map(int, input().split())) for _ in range(12))
print(solution(A))